HIP/hip-906.md

```diff
@@ -79,29 +77,66 @@ contract Assembly {
 }
 ```
 
-Once the above functionality has been implemented in services, an EOA or contract with hbars will be able to call the `hbarAllowance` and `hbarApprove` functions as follows:
+Once the above functionality has been implemented in services, an EOA with hbars will be able to call the `hbarAllowance` and `hbarApprove` functions as follows:
+
+```
+IHRC632(accountAddress).hbarAllowance(address spender)
+IHRC632(accountAddress).hbarApprove(address spender, int256 amount)
+```
+> The `hbarApprove` function will be callable by an EOA as long as the signatures required by the security model are met.
+
+## Acting on a Contract
+The `IHederaAccountService` system contract will be updated to add the `hbarAllowance` and `hbarApprove` functions.  
+The signature of these calls will be similar to the EOA calls except that they will include the contract address as the first argument.
+
+The following table describes the function selectors for the new `hbarAllownace` and `hbarApprove` functions and the associated function signatures and responses in `IHederaAccountService`.
+
+| Function Selector Hash | Function Signature                                         | Response               |                                                                                                      | 
+|------------------------|------------------------------------------------------------|------------------------|------------------------------------------------------------------------------------------------------|
+| 0xfec46666             | hbarAllowance(address owner, address spender)              | (ResponseCode, int256) | The response code from the call and the amount of hbar allowances currently available to the spender | 
+| 0xa0918464             | hbarApprove(address owner, address spender, int256 amount) | ResponseCode           | The response code from the call                                                                      |
+
+The solidity interface for `IHederaAccountService` will be the following after this HIP is implemented:
 
 ```
-IHRC(accountAddress).hbarAllowance(address spender)
-IHRC(accountAddress).hbarApprove(address spender, uint256 amount)
+interface IHederaAccountService {
+    ...
+    function hbarAllowance(address owner, address spender) external returns (responseCode, int256);
+    function hbarApprove(address owner, address spender, int256 amount) external returns (responseCode);
+    ...
+}
 ```
-> The `hbarApprove` function will be callable by either an EOA or a contract as long as the signatures required by the security model are met.
+
+Once the above functionality has been implemented in services, a contract with hbars will be able to call the `hbarAllowance` and `hbarApprove` functions as follows:
+
+```
+SampleContract is HederaAccountService {
+    function someFunction() {
+        ...    
+        this.hbarAllowance(address owner, address spender)
+        this.hbarApprove(address owner, address spender, int256 amount)
```

Comment:
The problem here (and I just notice at L99-100 above) is that these aren't calls nor signatures!  A call would look like

```
address owner = ...;
address spender = ...;
this.hbarApprove(owner, spender);
```

IOW, what's currently here is a combination of calling syntax (the `this.hbarApprove(...)` part) and declaration syntax (the `(address owner, address spender)` part).

Reply:
good point.   I will remove the type and just have the reader infer the initiation of the input parameters.